### ISSUE TITLE
Suggestion: Implement A Developer System / Branch / Channel

### ISSUE THREAD

**Issue by SemperFiWS** [cubecart/v6#1615]
With recent CC6 releases, new functionality has been introduced or existing functionality improved upon, which is great.
But unfortunately it has often contained bugs and a quick follow up release has been required to address them.

As a result you end up with live stores (a) being affected by these bugs and (b) needing to do multiple upgrade in quick succession.

One way to improve on the current system, would be to establish and implement a developer system, or branch or channel, whatever you want to call it.
i.e.
- have the current stable branch (e.g. 6.1.7)
- have a separate developer branch (e.g. 6.1.8)
- add a new setting to the store config (stable or development)
- if the store owner has that setting to development, they get notified about and can upgrade to the developer branch
- if the store owner has that setting to stable, they get notified about and can upgrade to the stable branch

This new setting could be used in live stores.
Or store owners could choose to have separate development stores.

Then instead of rushing the stable release, first do a developer release.

That way anyone who wants to can have a development store, they get get notified at that (but not in a live store) a new version is available.
They then update their development store and test the new version thoroughly, providing feedback, revisions get made as required.

Once a set period of time passes, and all is confirmed as working 100% correctly, the development version then becomes the current version, which live store owners can then use.

**Comment by briansandall** [comment]
Should have read this before commenting on the previous issue. ;)

Agree 100%. Git makes branching easy - let's use it.

**Comment by Dirty-Butter** [comment]
Sounds like a great idea to me! I would like to be included in such a branch, even though I'm not that knowledgeable. But I'm good for spotting spelling errors and directions that are unclear to the ordinary user. And sometimes I can suggest tweaks that turn out to be good ideas.

**Comment by abrookbanks** [comment]
Guys please don't think I'm ignoring you by closing thee issues. I'm not. I think we need change. 

The biggest concern I have is adoption for beta like branches. Version 5 was in beta for months. On final release a tonne of bugs came out of the works and like a lot of past releases it wasn't ready. Getting people to test software and report issues has historically proven to be tough. On those grounds these days I'm more inclined to push released out and fix as we find anything. 

I don't actually think the last few releases have been as buggy as they may have appeared. The biggest issue has been done to failed upgrades stores stuck between versions.

**Comment by Dirty-Butter** [comment]
> I'm more inclined to push released out and fix as we find anything.

I respectfully disagree with your reasoning. This approach effectively means you force all CC users to be beta testers. Store owner frustrations result in people leaving CC for other software choices.

**Comment by abrookbanks** [comment]
It needs to be marketed as something other than beta then. I agree with what you are saying but any what I would call as blockers get resolved fast with a new release. 

My mind ponders...

**Comment by briansandall** [comment]
I agree with @Dirty-Butter, but at the same time there is likely never going to be a release (beta or not) without bugs of some kind or another, so I can see an argument for both sides.

However, many store owners simply want a functioning eCommerce solution. They would rather run outdated software with major security flaws than face or lose an unsatisfied customer due to even potentially minor issues with their online purchasing experience.

I think it would be extremely beneficial to have a main, stable branch that can receive critical patches at any time without affecting or being affected by the next scheduled release version. The next release can still behave as releases do right now and stores will be able to stay secure without having to adopt a version that was released early (read: not ready for release) just to get a security fix.

**Comment by abrookbanks** [comment]
@briansandall I think that makes absolute sense.

**Comment by SemperFiWS** [comment]
As I commented in #1611 great to see this has been implemented :)

Quick question:
Will the functionality to select which branch a store runs / is notified about be adding as a setting?

**Comment by abrookbanks** [comment]
Stores will only be able to run from the master branch if updated via the control panel of CubeCart. We could implement a system to allow the merchant to choose which branch to operate from depending on their stability preferences.

**Comment by Dirty-Butter** [comment]
Hands-off owners will definitely want stable CC upgraded from Admin. Code-comfortable will like to be able to choose upgrading with bells and whistles, even if it means manual upgrade.

**Comment by abrookbanks** [comment]
I've just completely cocked the branches up. Oh well.. I took the preview release form Master.. I'll get it straightened up next time.
